Basic docs (#90)
leancodepl/contractsgenerator: docs/tutorial.md

# Tutorial

This tutorial will show you how to write your first command, query and operation.

## Basics

Contracts are written in C#. You can use any C# feature, provided that the feature does not introduce any "logic". Any
non-structural code in contracts will be either rejected or ignored. You can find more about guidelines what features
can be used [here](./guidelines.md).

Contracts are based on 5 basic types:

1. DTOs (Data Transfer Objects),
2. Enums (also a form of Data Transfer Objects),
3. Queries,
4. Commands, and
5. Operations.

> The first two (DTOs & Enums) represent plain data. Latter three (Queries, Commands & Operations) represent API endpoints

Reviewer: ```suggestion
The first two (DTOs & Enums) represent plain data. The last three (Queries, Commands & Operations) represent API endpoints
```

https://grammarly.com/blog/former-vs-latter
> Do not use former or latter when you are writing about more than two things.

DTO is used to pass data to or return data from the backend. It carries no other meaning. You can use almost all
[known types](./types.md) (with exceptions) or types that are defined by your project or referenced project (provided
that the referenced project carries the contracts payload) as properties.

A DTO can also define any number of constants, it can extend any DTO class and implement any number of interfaces. It
can also have any number of attributes assigned (they don't carry meaning on contracts level but might be used by
client or server implementations).

Here is an example DTO:

```csharp

// This will result in a DTO `IInterfaceDTO` with a single string property - `A`.
public interface IInterfaceDTO
{
    public string A { get; set; }
}

// This will result in a DTO with one property - `B` - that extends `IInterfaceDTO` type.
// The `abstract` modifier will be ignored by the generator.
public abstract class BaseDTO : IInterfaceDTO
{
    // Although it is implemented here, it will not be considered internal
    // `BaseDTO` property - it will come from extension from `IInterfaceDTO`
    public string A { get; set; }

    public int B { get; set; }
}

> // This will result in a DTO with one property - `C` -  and a constant `MyConstant` that extends `BaseDTO`.

Reviewer: ```suggestion
// This will result in a DTO that extends `BaseDTO` with one property - `C` -  and a constant `MyConstant`.
```

result, the interface is not generic - you don't specify return type, contrary to queries.

Commands should be named as imperative, i.e. `rate order` or `upload photo`.

Command validation is supported directly by the contracts. The way validation is reported is by the usage of error codes.
The codes are specified as a part of the command class, in a special inner class named `ErrorCodes`. An error code is
basically a numeric constant that server can use the report the error and client can use to distinguish it from different
errors.

```csharp
public class RateOrder : ICommand
{
    public Guid Id { get; set; }
    public int Rating { get; set; }

    // The static here is ignored by the contracts. Only the constants in `ErrorCodes` class are used.
    public static class ErrorCodes
    {
        public const OrderDoesNotExist = 1;
        public const OrderAlreadyRated = 2;
        public const RatingIsOutOfRange = 3;
    }
}
```

### Operations

Operations are an exception to the CQRS rule. Sometimes there are cases (esp. when integrating with external services)
where you can't split _reading_ the data from _changing_ it. You can think of operations as of commands that return

> a result. Operations are not validated, contrary to commands (you need to model validation yourself).

Reviewer: ```suggestion
results. Operations are not validated, contrary to commands (you need to model validation yourself).
```

> A common example for operation is payment - most of the payment processors have "pay" action that both does the payment

Reviewer: ```suggestion
A common example of an operation is payment - most of the payment processors have "pay" action that both does the payment
```

> and return some metadata about the payment.

Reviewer: ```suggestion
and returns some metadata about the payment.
```
